sdk/program/src/message/versions.rs

```diff
@@ -0,0 +1,263 @@
+            Self::V0(message) => message.sanitize(),
+        }
+    }
+}
+
+impl Serialize for MessageVersions {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        match self {
+            Self::Original(message) => {
+                let mut seq = serializer.serialize_tuple(1)?;
+                seq.serialize_element(message)?;
+                seq.end()
+            }
+            Self::V0(message) => {
+                let mut seq = serializer.serialize_tuple(2)?;
+                seq.serialize_element(&VERSION_PREFIX)?;
+                seq.serialize_element(message)?;
+                seq.end()
+            }
+        }
+    }
+}
+
+enum MessagePrefix {
+    Original(u8),
+    Versioned(u8),
+}
+
+impl<'de> Deserialize<'de> for MessagePrefix {
+    fn deserialize<D>(deserializer: D) -> Result<MessagePrefix, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        struct PrefixVisitor;
+
+        impl<'de> Visitor<'de> for PrefixVisitor {
+            type Value = MessagePrefix;
+
+            fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
+                formatter.write_str("message prefix byte")
+            }
+
+            fn visit_u8<E>(self, byte: u8) -> Result<MessagePrefix, E> {
+                if byte & VERSION_PREFIX != 0 {
+                    Ok(MessagePrefix::Versioned(byte ^ VERSION_PREFIX))
+                } else {
+                    Ok(MessagePrefix::Original(byte))
+                }
+            }
+        }
+
+        deserializer.deserialize_u8(PrefixVisitor)
+    }
+}
+
+impl<'de> Deserialize<'de> for MessageVersions {
+    fn deserialize<D>(deserializer: D) -> Result<MessageVersions, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        struct MessageVisitor;
+
+        impl<'de> Visitor<'de> for MessageVisitor {
+            type Value = MessageVersions;
+
+            fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
+                formatter.write_str("message bytes")
+            }
+
+            fn visit_seq<A>(self, mut seq: A) -> Result<MessageVersions, A::Error>
+            where
+                A: SeqAccess<'de>,
+            {
+                let prefix: MessagePrefix = seq
+                    .next_element()?
+                    .ok_or_else(|| de::Error::invalid_length(0, &self))?;
+
+                match prefix {
+                    MessagePrefix::Original(num_required_signatures) => {
+                        // The remaining fields of the original Message struct after the first byte.
+                        #[derive(Serialize, Deserialize)]
+                        struct RemainingMessage {
+                            pub num_readonly_signed_accounts: u8,
+                            pub num_readonly_unsigned_accounts: u8,
+                            #[serde(with = "short_vec")]
+                            pub account_keys: Vec<Pubkey>,
```

Comment:
the duplication here is unfortunate

Reply:
Agreed, I spent a fair bit of time trying to "peek" at the first element to avoid this but couldn't get to a good solution. I'll add a comment in the `Message` struct pointing to this code.